mox/impl/src/lib.rs

```diff
@@ -1,245 +1,239 @@
+use proc_macro2::{Span, TokenStream};
 use quote::{quote, ToTokens};
-use snax::{ParseError, SnaxAttribute, SnaxFragment, SnaxItem, SnaxSelfClosingTag, SnaxTag};
-use std::borrow::Cow;
+use std::convert::TryFrom;
+use syn::{parse::Parse, parse_macro_input, spanned::Spanned};
 
-#[proc_macro_error(allow_not_macro)]
-#[proc_macro_hack]
+#[proc_macro]
 pub fn mox(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
-    let item = snax::parse(input.into()).map_err(Error::SnaxError).unwrap_or_abort();
-    let item = MoxItem::from(item);
+    let item = parse_macro_input!(input as MoxItem);
     quote!(#item).into()
 }
 
 enum MoxItem {
     Tag(MoxTag),
-    TagNoChildren(MoxTagNoChildren),
-    Fragment(Vec<MoxItem>),
-    Content { span: Span, stream: TokenStream },
+    Expr(MoxExpr),
 }
 
-impl MoxItem {
-    fn span(&self) -> Span {
-        match self {
-            MoxItem::Tag(tag) => tag.span(),
-            MoxItem::TagNoChildren(tag) => tag.span(),
-            MoxItem::Content { span, .. } => *span,
-            MoxItem::Fragment(_children) => todo!(),
-        }
-    }
+struct MoxTag {
+    name: syn::ExprPath,
+    attributes: Vec<MoxAttr>,
+    children: Vec<MoxItem>,
 }
 
-impl From<SnaxItem> for MoxItem {
-    fn from(item: SnaxItem) -> Self {
-        match item {
-            SnaxItem::Tag(t) => MoxItem::Tag(MoxTag::from(t)),
-            SnaxItem::SelfClosingTag(t) => MoxItem::TagNoChildren(MoxTagNoChildren::from(t)),
-            SnaxItem::Fragment(SnaxFragment { children }) => {
-                MoxItem::Fragment(children.into_iter().map(MoxItem::from).collect())
-            }
-            SnaxItem::Content(atom) => {
-                MoxItem::Content { span: atom.span(), stream: wrap_content_tokens(atom) }
-            }
-        }
-    }
+struct MoxAttr {
+    name: syn::Ident,
+    value: Option<syn::Expr>,
 }
 
-fn wrap_content_tokens(tt: TokenTree) -> TokenStream {
-    let mut new_stream = quote!(#tt);
-    match tt {
-        TokenTree::Group(g) => {
-            let mut tokens = g.stream().into_iter();
-            if let Some(TokenTree::Punct(p)) = tokens.next() {
-                if p.as_char() == '%' {
-                    // strip the percent sign off the front
-                    new_stream = TokenStream::new();
-                    new_stream.extend(tokens);
-                    new_stream = quote!(text(format!(#new_stream)));
-                }
-            }
-        }
-        tt @ TokenTree::Ident(_) | tt @ TokenTree::Literal(_) => {
-            new_stream = quote!(text(#tt));
-        }
-        TokenTree::Punct(p) => emit_error!(p.span(), "'{}' not valid in item position", p),
+struct MoxExpr {
+    expr: syn::Expr,
+}
+
+impl Parse for MoxItem {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let parse_config = syn_rsx::ParserConfig::new()
+            .number_of_top_level_nodes(1)
+            .type_of_top_level_nodes(syn_rsx::NodeType::Element);
```

Comment:
Does syn-rsx not support multiple kinds of root values? If this is true, should we `impl Parse for MoxTag` instead of `MoxItem`?

Reply:
There will be support for the fragments soon with a new [NodeType::Fragment](https://docs.rs/syn-rsx/0.8.0-beta.1/syn_rsx/enum.NodeType.html#variant.Fragment). This means we should drop restriction "top node as a tag" and allow anything which results into the value. This is also may be useful for a code generation.

Reply:
Right, but the `type_of_top_level_nodes(Element)` call here makes me think that we're restricted to a single top level node type. Am I misunderstanding?

Reply:
By default there's no restriction on amount or type of top level nodes. `type_of_top_level_nodes` and `number_of_top_level_nodes` allow to configure restrictions.

Reply:
Ah, good to know! So maybe what we should do is remove the `type_of_top_level_nodes` call and leave the `number_of_top_level_nodes` call.